twine/commands/upload.py

```diff
@@ -140,6 +141,9 @@ def upload(upload_settings: settings.Settings, dists: List[str]) -> None:
 
         resp = repository.upload(package)
 
+        if upload_settings.verbose:
+            logger.info(dump.dump_all(resp).decode("utf-8"))
```

Comment:
Sample output:

```
Uploading example_pkg_bhrutledge-0.0.6.post1-py3-none-any.whl
< POST /legacy/ HTTP/1.1
< Host: test.pypi.org
< User-Agent: twine/3.7.2.dev5+g2cf7138.d20220119 importlib_metadata/4.10.1 pkginfo/1.8.2 requests/2.27.1 requests-toolbelt/0.9.1 tqdm/4.62.3 CPython/3.9.10
< Accept-Encoding: gzip, deflate
< Accept: */*
< Connection: keep-alive
< Content-Type: multipart/form-data; boundary=2c0ed756d4054e4cb4a0544670ff12b0
< Content-Length: 6702
< Authorization: Basic X190b2tlbl9fOnB5cGktQWdFTmRHVnpkQzV3ZVhCcExtOXlad0lrWW1JelpEWXlaRFF0TnpNek15MDBNakV3TFdKalpqUXRNell6T1RZeU1tSmhPR0l4QUFJbGV5SndaWEp0YVhOemFXOXVjeUk2SUNKMWMyVnlJaXdnSW5abGNuTnBiMjRpT2lBeGZRQUFCaUFXQnAzNnl1eHlBSEJQbWhyV1hXazVIMlBDVXNNNkdYcXpIU1dwUmpyTml3
< 
<< Request body is not a string-like type >>
> HTTP/1.1 200 OK
> Connection: keep-alive
> Content-Length: 0
> Content-Security-Policy: base-uri 'self'; block-all-mixed-content; connect-src 'self' https://api.github.com/repos/ fastly-insights.com *.fastly-insights.com *.ethicalads.io https://api.pwnedpasswords.com; default-src 'none'; font-src 'self' fonts.gstatic.com; form-action 'self'; frame-ancestors 'none'; frame-src 'none'; img-src 'self' https://warehouse-test-camo.ingress.cmh1.psfhosted.org/ www.google-analytics.com *.fastly-insights.com *.ethicalads.io; script-src 'self' www.googletagmanager.com www.google-analytics.com *.fastly-insights.com *.ethicalads.io 'sha256-U3hKDidudIaxBDEzwGJApJgPEf2mWk6cfMWghrAa6i0='; style-src 'self' fonts.googleapis.com *.ethicalads.io 'sha256-2YHqZokjiizkHi1Zt+6ar0XJ0OeEy/egBnlm+MDMtrM=' 'sha256-47DEQpj8HBSa+/TImW+5JCeuQeRkm5NMpJWZG3hSuFU='; worker-src *.fastly-insights.com
> Content-Type: text/html; charset=UTF-8
> Referrer-Policy: origin-when-cross-origin
> Server: nginx/1.13.9
> Accept-Ranges: bytes
> Date: Wed, 19 Jan 2022 00:36:35 GMT
> X-Served-By: cache-bos4659-BOS
> X-Cache: MISS
> X-Cache-Hits: 0
> X-Timer: S1642552595.060024,VS0,VE327
> Vary: Cookie, Authorization, Cookie
> Strict-Transport-Security: max-age=31536000; includeSubDomains; preload
> X-Frame-Options: deny
> X-XSS-Protection: 1; mode=block
> X-Content-Type-Options: nosniff
> X-Permitted-Cross-Domain-Policies: none
> X-Robots-Header: noindex
> 
```

Reply:
Do we really want that level of detail? I mean I wrote the utility to dump stuff, so I'm not against using it. I'm just wondering if this is too much and if a non-PyPI server with lots of redirects could make this output borderline useless

Reply:
I wondered that, too. If I follow, [`dump_response`](https://toolbelt.readthedocs.io/en/latest/dumputils.html#requests_toolbelt.utils.dump.dump_response) would show the same level of detail, but without redirects.

I think the point of this is to help folks debug configuration issues and unexpected responses. Below is another example, this time of an error response. In both cases, I think the amount of info might mask what's useful. I could try to arrive come up with a middle ground, but then again, who knows what's useful in any specific situation?

```
Uploading example_pkg_bhrutledge-0.0.5-py3-none-any.whl
< POST /legacy/ HTTP/1.1
< Host: test.pypi.org
< User-Agent: twine/3.7.2.dev5+g2cf7138.d20220119 importlib_metadata/4.10.1 pkginfo/1.8.2 requests/2.27.1 requests-toolbelt/0.9.1 tqdm/4.62.3 CPython/3.9.10
< Accept-Encoding: gzip, deflate
< Accept: */*
< Connection: keep-alive
< Content-Type: multipart/form-data; boundary=73cc01a8b2244236a22f87dc2a2a4eb3
< Content-Length: 6620
< Authorization: Basic X190b2tlbl9fOnB5cGktQWdFTmRHVnpkQzV3ZVhCcExtOXlad0lrWW1JelpEWXlaRFF0TnpNek15MDBNakV3TFdKalpqUXRNell6T1RZeU1tSmhPR0l4QUFJbGV5SndaWEp0YVhOemFXOXVjeUk2SUNKMWMyVnlJaXdnSW5abGNuTnBiMjRpT2lBeGZRQUFCaUFXQnAzNnl1eHlBSEJQbWhyV1hXazVIMlBDVXNNNkdYcXpIU1dwUmpyTml3
< 
<< Request body is not a string-like type >>
> HTTP/1.1 400 File already exists. See https://test.pypi.org/help/#file-name-reuse for more information.
> Connection: keep-alive
> Content-Length: 469
> Content-Security-Policy: base-uri 'self'; block-all-mixed-content; connect-src 'self' https://api.github.com/repos/ fastly-insights.com *.fastly-insights.com *.ethicalads.io https://api.pwnedpasswords.com; default-src 'none'; font-src 'self' fonts.gstatic.com; form-action 'self'; frame-ancestors 'none'; frame-src 'none'; img-src 'self' https://warehouse-test-camo.ingress.cmh1.psfhosted.org/ www.google-analytics.com *.fastly-insights.com *.ethicalads.io; script-src 'self' www.googletagmanager.com www.google-analytics.com *.fastly-insights.com *.ethicalads.io 'sha256-U3hKDidudIaxBDEzwGJApJgPEf2mWk6cfMWghrAa6i0='; style-src 'self' fonts.googleapis.com *.ethicalads.io 'sha256-2YHqZokjiizkHi1Zt+6ar0XJ0OeEy/egBnlm+MDMtrM=' 'sha256-47DEQpj8HBSa+/TImW+5JCeuQeRkm5NMpJWZG3hSuFU='; worker-src *.fastly-insights.com
> Content-Type: text/html; charset=UTF-8
> Referrer-Policy: origin-when-cross-origin
> Server: nginx/1.13.9
> Accept-Ranges: bytes
> Date: Wed, 19 Jan 2022 14:01:51 GMT
> X-Served-By: cache-bos4648-BOS
> X-Cache: MISS
> X-Cache-Hits: 0
> X-Timer: S1642600911.339231,VS0,VE163
> Vary: Cookie, Cookie, Authorization
> Strict-Transport-Security: max-age=31536000; includeSubDomains; preload
> X-Frame-Options: deny
> X-XSS-Protection: 1; mode=block
> X-Content-Type-Options: nosniff
> X-Permitted-Cross-Domain-Policies: none
> X-Robots-Header: noindex
> 
<html>
 <head>
  <title>400 File already exists. See https://test.pypi.org/help/#file-name-reuse for more information.</title>
 </head>
 <body>
  <h1>400 File already exists. See https://test.pypi.org/help/#file-name-reuse for more information.</h1>
  The server could not comply with the request since it is either malformed or otherwise incorrect.<br/><br/>
File already exists. See https://test.pypi.org/help/#file-name-reuse for more information.


 </body>
</html>
HTTPError: 400 Bad Request from https://test.pypi.org/legacy/
File already exists. See https://test.pypi.org/help/#file-name-reuse for more information.
```

Reply:
I'm not trying to say it's bad or wrong. I do wonder, once again if we had a repeatable verbose option like any good CLI if we could decide what amount of information to log. If it's just one level, we log short bits of info about each redirect (status and request url) and then the final response info. If we have two levels we log this for the single response. Three levels means we log everything with dumpall

We've been inexplicably opposed to that for vague notions of complexity that aren't real based on extensive experience of my own. But we now have to pick something to represent here

Reply:
I'm agree with you.
It's difficult to know what is relevant in terms of log.

I think the best will be to have several verbose options.
For exemple -v for limited log and -vvv with the content of all headers of HTTP request.
Or an environment variable with a number for the log level details.

Reply:
@bhrutledge Just a heads up that posting your `Authorization` header here has the effect of leaking your upload token:

```python
>>> import base64
>>> s = """X190b2tlbl9fOnB5cGktQWdFTmRHVnpkQzV3ZVhCcExtOXlad0lrWW1JelpEWXlaRFF0TnpNek15MDBNakV3TFdKalpqUXRNell6T1RZeU1tSmhPR0l4QUFJbGV5SndaWEp0YVhOemFXOXVjeUk2SUNKMWMyVnlJaXdnSW5abGNuTnBiMjRpT2lBeGZRQUFCaUFXQnAzNnl1eHlBSEJQbWhyV1hXazVIMlBDVXNNNkdYcXpIU1dwUmpyTml3"""
>>> base64.b64decode(s)
b'__token__:pypi-AgENdGVzdC5weXBpLm9yZwIkYmIzZDYyZDQtNzMzMy00MjEwLWJjZjQtMzYzOTYyMmJhOGIxAAIleyJwZXJtaXNzaW9ucyI6ICJ1c2VyIiwgInZlcnNpb24iOiAxfQAABiAWBp36yuxyAHBPmhrWXWk5H2PCUsM6GXqzHSWpRjrNiw'
```

This token has now been revoked from TestPyPI. We might want to rethink this feature, or at least ensure that we aren't providing a mechanism to leak credentials in CI logs or  issues like this.

Reply:
> I do wonder, once again if we had a repeatable verbose option like any good CLI if we could decide what amount of information to log. We've been inexplicably opposed to that for vague notions of complexity that aren't real based on extensive experience of my own.

@sigmavirus24 Using multiple levels of verbosity was discussed in https://github.com/pypa/twine/pull/669#discussion_r446508397 and https://github.com/pypa/twine/pull/815#issuecomment-932902679, and I don't think the reason for postponing was a vague notion. I'm also not opposed to it, but am keen to make `-v` as useful as possible.

Reply:
> Just a heads up that posting your `Authorization` header here has the effect of leaking your upload token:

@di Yikes! Thanks for taking care of it on TestPyPI.

It's clear that `dump_all` and `dump_response` is too verbose for `-v`, but seeing the output is a useful guide for what to include. I'll follow up with another attempt.